metals/src/main/scala/scala/meta/internal/metals/codeactions/AddingBracesCodeAction.scala

```diff
@@ -0,0 +1,641 @@
+                  ),
+                  distanceToThen
+                ),
+                (
+                  createCodeActionForPotentialBlockHolder(
+                    termIf,
+                    path,
+                    document,
+                    termIf.elsep,
+                    "else expression"
+                  ),
+                  distanceToElseP
+                )
+              ).minBy(_._2)
+            ).map(_._1).toOption.flatten.toList
+
+          case termFor: Term.For =>
+            createCodeActionForPotentialBlockHolder(
+              termFor,
+              path,
+              document,
+              termFor.body,
+              "for expression"
+            ).toSeq
+          case termForYield: Term.ForYield =>
+            createCodeActionForPotentialBlockHolder(
+              termForYield,
+              path,
+              document,
+              termForYield.body,
+              "yield expression"
+            ).toSeq
+          case termWhile: Term.While =>
+            createCodeActionForPotentialBlockHolder(
+              termWhile,
+              path,
+              document,
+              termWhile.body,
+              "while expression",
+              "do"
+            ).toSeq
+          case _: Defn.GivenAlias => None.toSeq
+          case template: Template =>
+            val title = template.parent
+              .collectFirst {
+                case _: Defn.Enum => "enum definition"
+                case _: Defn.Trait => "trait definition"
+                case _: Defn.Object => "object definition"
+                case _: Defn.Class => "class definition"
+              }
+              .getOrElse("template")
+            template.parent.flatMap {
+              createCodeActionForTemplateHolder(
+                _,
+                path,
+                document,
+                template,
+                title
+              )
+            }.toSeq
+          case termBlock: Term.Block if !termBlock.parent.exists(_ match {
+                case _: Term.Apply => true
+                case _ => false
+              }) =>
+            termBlock.parent.flatMap { blockHolder =>
+              createCodeActionForPotentialBlockHolder(
+                blockHolder,
+                path,
+                document,
+                termBlock,
+                "block"
+              )
+            }
+          case termMatch: Term.Match =>
+            createCodeActionForTermMatch(
+              termMatch,
+              path,
+              document,
+              "match cases"
+            )
+          case _ => None
+        }
+      }
+    result.toSeq.flatten
+  }
+
+  private def hasBraces(template: Template): Boolean = {
+    util
+      .Try(template.stats.maxBy(_.pos.end).pos.end)
+      .getOrElse(-1) != template.pos.end && util
+      .Try(template.tokens.maxBy(_.pos.end).text)
+      .getOrElse("") == "}"
+  }
+
+  private def hasBraces(termBlock: Term.Block): Boolean = {
```

Comment:
```suggestion
  private def hasBraces(term: Tree, stats: Seq[Tree]): Boolean = {
```
we could replace the two functions with this since they are identical.

Reply:
it's a private function so no need for that. It's more important when designing a library API, but here others can stilll change anything and we will be able to catch that on a review.

Reply:
hmmm… I see. no problem.

Reply:
oh, oops! the `.stats` member of `termBlock` and `template` are used in the body of this method. But this member is unavailable on `Tree` type.

Reply:
I am going to replace the method with `isStatsHolderBraced(tree: Tree, stats: List[Tree])`

Reply:
I think the way for reducing the number of calls to `codeActionForBlockHolder` is to extract the `(blockHolder, blockEmbraceable)` pair of various blockHolders and then pass the result to a single call to `codeActionForBlockHolder`.